### PR TITLE
Handle ALTER Index Queries.

### PR DESCRIPTION
**Pull request by vamsi-amazon**
### Description
* ALTER INDEX syntax updates.
* Handle ALTER INDEX Queries with auto_refresh=false in sql plugin as it requires stopping spark job.
* Sending rest of the ALTER INDEX queries are streaming queries in separate job to spark instance.
* Refactored existing drop operation.
* Introduced rolling back in flint index operations.
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by seankao-az — DISMISSED** [review]

File: spark/src/main/java/org/opensearch/sql/spark/flint/model/FlintIndexDetailsRequest.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    public Builder dataSourceName(String dataSourceName) {
+      this.dataSourceName = dataSourceName;
+      return this;
+    }
+
+    public FlintIndexDetailsRequest build() {
+      if (indexPattern == null || indexPattern.isEmpty()) {
+        throw new IllegalArgumentException("indexPattern cannot be null or empty");
+      }
+      if (isIndexStateRequired && (dataSourceName == null || dataSourceName.isEmpty())) {
```
Comment (penghuo) [inline spark/src/main/java/org/opensearch/sql/spark/flint/model/FlintIndexDetailsRequest.java:49]: nit, dataSourceName == null || dataSourceName.isEmpty() -> Strings.isEmptyOrNull()

File: spark/src/main/java/org/opensearch/sql/spark/flint/FlintIndexMetadata.java [hunk 2]
```diff
@@ -5,42 +5,22 @@
...
-  public static final String AUTO_REFRESH_DEFAULT = "false";
-
-  public static final String APP_ID = "SERVERLESS_EMR_VIRTUAL_CLUSTER_ID";
-  public static final String FLINT_INDEX_STATE_DOC_ID = "latestId";
 
   private final String jobId;
   private final boolean autoRefresh;
   private final String appId;
   private final String latestId;
+  private final FlintIndexStateModel indexStateModel;
```
Comment (penghuo) [inline spark/src/main/java/org/opensearch/sql/spark/flint/FlintIndexMetadata.java:20]: question? in case static method `getFlintIndexState(stateStore, datasourceName).apply(latestId.get());` can fetch IndexState, what is the major reason of code refactor?

Comment (vamsi-amazon) [reply spark/src/main/java/org/opensearch/sql/spark/flint/FlintIndexMetadata.java:20]: I was viewing it from cron job to clean all the indices of a datasource.
I was thinking to get IndexStateModel along with the IndexMetadata and this would be helpful in filtering out indices based on state.
Now I am implementing ALTER index and will probably remove this. Converting to draft for now to include alter index implementation as well.

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java [hunk 3]
```diff
@@ -1238,20 +1238,17 @@ public void testWeekOfYearWithTimeType() {
                 DSL.week(
                     functionProperties, DSL.literal(new ExprTimeValue("12:23:34")), DSL.literal(0)),
                 "week(TIME '12:23:34', 0)",
-                LocalDate.now(functionProperties.getQueryStartClock()).get(ALIGNED_WEEK_OF_YEAR)
```
Comment (vamsi-amazon) [inline core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:1241]: These changes can be pulled into another PR. Added just to unblock the build of current PR.

File: spark/src/main/java/org/opensearch/sql/spark/flint/operation/FlintIndexOpAlter.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      FlintIndexMetadataService flintIndexMetadataService) {
+    super(stateStore, datasourceName);
+    this.emrServerlessClient = emrServerlessClient;
+    this.flintIndexMetadataService = flintIndexMetadataService;
+    this.flintIndexOptions = flintIndexOptions;
+  }
+
+  @Override
+  protected boolean validate(FlintIndexState state) {
+    return state == FlintIndexState.ACTIVE || state == FlintIndexState.REFRESHING;
```
Comment (penghuo) [inline spark/src/main/java/org/opensearch/sql/spark/flint/operation/FlintIndexOpAlter.java:43]: FlintIndexState.ACTIVE is validate state?

File: spark/src/main/java/org/opensearch/sql/spark/flint/operation/FlintIndexOpAlter.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  FlintIndexState transitioningState() {
+    return FlintIndexState.UPDATING;
+  }
+
+  @SneakyThrows
+  @Override
+  void runOp(FlintIndexMetadata flintIndexMetadata, FlintIndexStateModel flintIndexStateModel) {
+    LOG.debug(
+        "Running alter index operation for index: {}", flintIndexMetadata.getOpensearchIndexName());
+    this.flintIndexMetadataService.updateIndexToManualRefresh(
```
Comment (penghuo) [inline spark/src/main/java/org/opensearch/sql/spark/flint/operation/FlintIndexOpAlter.java:56]: if auto_refresh is changed to false, but cancelStreamingJob failed, does ALTER INDEX auto_refresh=false retry able?

File: spark/src/test/java/org/opensearch/sql/spark/flint/FlintIndexMetadataServiceSpecTest.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+  @Test
+  public void testGetFlintIndexMetadataWithIndexState() {
+
+    skippingIndex.createIndex();
+    skippingIndexState.refreshing();
+    coveringIndex.createIndex();
+    coveringIndexState.refreshing();
+    mvIndex.createIndex();
+    mvIndexState.active();
```
Comment (penghuo) [inline spark/src/test/java/org/opensearch/sql/spark/flint/FlintIndexMetadataServiceSpecTest.java:66]: why MV state is active, and other indices state is refreshing?

File: spark/src/test/java/org/opensearch/sql/spark/flint/FlintIndexMetadataServiceSpecTest.java [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+      coveringIndexState.deleted();
+      mvIndex.deleteIndex();
+      mvIndexState.deleted();
+    }
+  }
+
+  @Test
+  public void testUpdateMapping() {
+    skippingIndex.createIndex();
+    LOG.fatal(
```
Comment (penghuo) [inline spark/src/test/java/org/opensearch/sql/spark/flint/FlintIndexMetadataServiceSpecTest.java:123]: remove log?

File: spark/src/test/java/org/opensearch/sql/spark/flint/FlintIndexMetadataServiceSpecTest.java [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+      skippingIndexState.deleted();
+      coveringIndex.deleteIndex();
+      coveringIndexState.deleted();
+      mvIndex.deleteIndex();
+      mvIndexState.deleted();
+    }
+  }
+
+  @Test
+  public void testUpdateMapping() {
```
Comment (penghuo) [inline spark/src/test/java/org/opensearch/sql/spark/flint/FlintIndexMetadataServiceSpecTest.java:121]: no assert?

File: spark/src/test/java/org/opensearch/sql/spark/asyncquery/IndexQuerySpecTest.java [hunk 9]
```diff
@@ -368,9 +341,11 @@ public EMRServerlessClient getClient() {
               AsyncQueryExecutionResponse asyncQueryResults =
                   asyncQueryExecutorService.getAsyncQueryResults(response.getQueryId());
               assertEquals("FAILED", asyncQueryResults.getStatus());
-              assertEquals("cancel job timeout", asyncQueryResults.getError());
+              assertEquals(
+                  "Cancel job timeout for Application ID: mockAppId, Job ID: mockJobId",
```
Comment (penghuo) [inline spark/src/test/java/org/opensearch/sql/spark/asyncquery/IndexQuerySpecTest.java:345]: should appId and jobId are exposed?

File: spark/src/test/java/org/opensearch/sql/spark/asyncquery/IndexQuerySpecTest.java [hunk 10]
```diff
@@ -469,50 +427,35 @@ public EMRServerlessClient getClient() {
...
               // 2. fetch result
-              assertEquals(
-                  "SUCCESS",
-                  asyncQueryExecutorService
-                      .getAsyncQueryResults(response.getQueryId())
-                      .getStatus());
-
+              AsyncQueryExecutionResponse asyncQueryExecutionResponse =
+                  asyncQueryExecutorService.getAsyncQueryResults(response.getQueryId());
+              assertEquals("SUCCESS", asyncQueryExecutionResponse.getStatus());
```
Comment (penghuo) [inline spark/src/test/java/org/opensearch/sql/spark/asyncquery/IndexQuerySpecTest.java:432]: nit, what is the difference? any consideration?

File: spark/src/test/java/org/opensearch/sql/spark/asyncquery/IndexQuerySpecTest.java [hunk 11]
```diff
@@ -549,76 +492,55 @@ public GetJobRunResult getJobRunResult(String applicationId, String jobId) {
...
+              flintIndexJob.assertState(FlintIndexState.DELETING);
             });
   }
 
-  /**
-   * No Job running, expectation is
-   *
-   * <p>(1) not call EMR-S (2) change index state to: DELETED
-   */
+  /** No Job running, expectation is */
```
Comment (penghuo) [inline spark/src/test/java/org/opensearch/sql/spark/asyncquery/IndexQuerySpecTest.java:522]: why ignore emr-s call verification?

File: spark/src/main/java/org/opensearch/sql/spark/flint/operation/FlintIndexOpAlter.java (hunk 4 above)
Comment (vamsi-amazon) [reply spark/src/main/java/org/opensearch/sql/spark/flint/operation/FlintIndexOpAlter.java:43]: Before altering the index should be either active or refreshing? Do you see any issue with this?

File: spark/src/test/java/org/opensearch/sql/spark/asyncquery/IndexQuerySpecTest.java (hunk 9 above)
Comment (vamsi-amazon) [reply spark/src/test/java/org/opensearch/sql/spark/asyncquery/IndexQuerySpecTest.java:345]: Good callout. I forgot MS.

File: spark/src/test/java/org/opensearch/sql/spark/asyncquery/IndexQuerySpecTest.java (hunk 10 above)
Comment (vamsi-amazon) [reply spark/src/test/java/org/opensearch/sql/spark/asyncquery/IndexQuerySpecTest.java:432]: Aaah...changed as part of refactoring. I was storing response in a separate variable for other test cases to assert Error.

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/IndexDMLHandler.java [hunk 12]
```diff
@@ -88,10 +102,48 @@ public DispatchQueryResponse submit(
...
             System.currentTimeMillis() - startTime,
             System.currentTimeMillis());
-    String resultIndex = dataSourceMetadata.getResultIndex();
-    createIndexDMLResult(stateStore, resultIndex).apply(indexDMLResult);
+    createIndexDMLResult(stateStore, dataSourceMetadata.getResultIndex()).apply(indexDMLResult);
+    return asyncQueryId;
+  }
 
-    return new DispatchQueryResponse(asyncQueryId, DROP_INDEX_JOB_ID, resultIndex, null);
+  private void executeIndexOp(
```
Comment (dai-chen) [inline spark/src/main/java/org/opensearch/sql/spark/dispatcher/IndexDMLHandler.java:109]: I'm adding `VACUUM` operation and thinking we may want to make this abstract and have separate handler for Drop, Alter and Vacuum in future?

File: spark/src/main/java/org/opensearch/sql/spark/flint/operation/FlintIndexOpCancel.java [hunk 13]
```diff
@@ -20,57 +19,46 @@ public class FlintIndexOpCancel extends FlintIndexOp {
   private static final Logger LOG = LogManager.getLogger();
 
   private final EMRServerlessClient emrServerlessClient;
+  private final boolean isDropOperation;
```
Comment (dai-chen) [inline spark/src/main/java/org/opensearch/sql/spark/flint/operation/FlintIndexOpCancel.java:22]: any special reason to merge Drop and Cancel together if they may run individually? I think the previous design follows single responsibility principle?

File: spark/src/test/java/org/opensearch/sql/spark/asyncquery/IndexQuerySpecTest.java (hunk 11 above)
Comment (vamsi-amazon) [reply spark/src/test/java/org/opensearch/sql/spark/asyncquery/IndexQuerySpecTest.java:522]: So for refreshing, empty, active and creating states. we cancel the job and removed emrs job verification. Let me add call asserts.

For other transitioning states, the query will result in error and in that case we are asserting no emr calls.
Let me know your thoughts on this.

File: spark/src/main/java/org/opensearch/sql/spark/flint/operation/FlintIndexOpAlter.java (hunk 5 above)
Comment (vamsi-amazon) [reply spark/src/main/java/org/opensearch/sql/spark/flint/operation/FlintIndexOpAlter.java:56]: I am making the query retryable by allowing the auto_refresh=false even if existing state is already in auto_refresh=false. So any transient errors in cancelling job can be retried.

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/IndexDMLHandler.java (hunk 12 above)
Comment (vamsi-amazon) [reply spark/src/main/java/org/opensearch/sql/spark/dispatcher/IndexDMLHandler.java:109]: Yes, I was thinking to abstract out and then the changes are becoming huge. Considering the timeline and since the we have only two use cases I just didn't refactor.  But makes sense to abstract out these operations individually and bring common pieces to an abstract class.

File: spark/src/main/java/org/opensearch/sql/spark/flint/operation/FlintIndexOpCancel.java (hunk 13 above)
Comment (vamsi-amazon) [reply spark/src/main/java/org/opensearch/sql/spark/flint/operation/FlintIndexOpCancel.java:22]: Cancel operation is not an actual operation --> which is ideally alter and we already have a separate operation.

In case after discussions with peng, I better understand the cancel use case. I will have a separate class for it.

File: spark/src/main/java/org/opensearch/sql/spark/flint/operation/FlintIndexOpCancel.java [hunk 14]
```diff
@@ -28,7 +27,7 @@ public FlintIndexOpCancel(
   }
 
   public boolean validate(FlintIndexState state) {
-    return state == FlintIndexState.REFRESHING || state == FlintIndexState.CANCELLING;
+    return state == FlintIndexState.REFRESHING || state == FlintIndexState.ACTIVE;
```
Comment (vamsi-amazon) [inline spark/src/main/java/org/opensearch/sql/spark/flint/operation/FlintIndexOpCancel.java:30]: @penghuo  can we review this? why no `active` ?
Should we keep cancelling in this as it is a transitioning state and a job is already running to cancel?

**Comment by dai-chen** [comment]
This PR also incorporated latest Flint SQL grammar changes below, so there is no need to publish separate PR. @rupal-bq @seankao-az 

1. ALTER statement: https://github.com/opensearch-project/opensearch-spark/pull/279
2. BLOOM_FILTER skip type: https://github.com/opensearch-project/opensearch-spark/pull/283
3. ANALYZE statement: https://github.com/opensearch-project/opensearch-spark/pull/284

**Review by dai-chen — APPROVED** [review]

File: spark/src/main/java/org/opensearch/sql/spark/flint/operation/FlintIndexOpAlter.java (hunk 4 above)
Comment (penghuo) [reply spark/src/main/java/org/opensearch/sql/spark/flint/operation/FlintIndexOpAlter.java:43]: Any case index is active, but auto_refresh=false, only for retry case?

File: spark/src/main/java/org/opensearch/sql/spark/flint/operation/FlintIndexOpCancel.java (hunk 14 above)
Comment (vamsi-amazon) [reply spark/src/main/java/org/opensearch/sql/spark/flint/operation/FlintIndexOpCancel.java]: After discussing we have concluded that only when index state is in REFRESHING, the jobId in flint index metadata is correct. So, REFRESHING is the only viable initial state to cancel the operation.

**Review by dai-chen — APPROVED** [review]

**Review by penghuo — APPROVED** [review]
